internal/plugin/connectors/tcp/mssql/connector.go

```diff
@@ -0,0 +1,111 @@
+)
+
+// SingleUseConnector is used to create an authenticated connection to an MSSQL target
+type SingleUseConnector struct {
+	backendConn net.Conn
+	clientConn  net.Conn
+	logger      log.Logger
+}
+
+// Connect implements the tcp.Connector func signature
+//
+// It is the main method of the SingleUseConnector. It:
+//   1. Constructs connection details from the provided credentials map.
+//   2. Dials the backend using credentials.
+//   3. Runs through the connection phase steps to authenticate.
+//   4. Pipes all future bytes unaltered between client and server.
+//
+// Connect requires "user", "password", and "address" credentials.
+func (connector *SingleUseConnector) Connect(
+	clientConn net.Conn,
+	credentialValuesByID connector.CredentialValuesByID,
+) (net.Conn, error) {
+
+	connector.clientConn = clientConn
+
+	err := connector.ConnectToBackend(credentialValuesByID)
+	if err != nil {
+		connector.logger.Errorf("Failed to connect to mssql server: %s", err)
+		return nil, err
+	}
+
+	err = connector.performPreLoginHandshake()
+	if err != nil {
+		connector.logger.Errorf("Failed to handle client prelogin request: %s", err)
+		return nil, err
+	}
+
+	// Now that the backend connection is authenticated, send the client
+	// a successful authentication response.
+	if _, err = clientConn.Write(connector.CreateAuthenticationOKMessage()); err != nil {
+		return nil, err
+	}
+
+	return connector.backendConn, nil
+}
+
+// ConnectToBackend establishes the connection to the target database and sets
+// the backendConnection field.
+func (connector *SingleUseConnector) ConnectToBackend(credentialValuesByID connector.CredentialValuesByID) error {
+	var err error
+
+	err = validateCredentialsExist(credentialValuesByID)
+	if err != nil {
+		connector.logger.Errorf("Missing credentials", err)
+		return err
+	}
+
+	// Using DSN (Data Source Name) string because gomssql forces us to.
+	//
+	// NOTE: Secretless has some unfortunate naming collisions with the
+	// go-mssqldb driver package.  The driver package has its own concept of a
+	// "connector", and its connectors also have a "Connect" method.
+	driverConnector, err := mssql.NewConnector(
+		fmt.Sprintf(
+			"sqlserver://%s:%s@%s",
+			credentialValuesByID["user"],
+			credentialValuesByID["password"],
+			credentialValuesByID["address"],
+		),
+	)
+	if err != nil {
+		connector.logger.Errorf("Failed to create a go-mssqldb connector: %s", err)
+		return err
+	}
+
+	driverConn, err := driverConnector.Connect(context.Background())
+	if err != nil {
+		connector.logger.Errorf("failed to connect to mssql server: %s", err)
+		return err
+	}
+
+	// Verify the driverConn is an mssql driverConn object and get its underlying transport
+	mssqlConn := driverConn.(*mssql.Conn)
+	connector.backendConn = mssqlConn.NetConn()
+
+	return nil
+}
+
+func validateCredentialsExist(credentialValuesByID connector.CredentialValuesByID) error {
+	requiredCredentials := []string{"user", "password", "address"}
```

Comment:
can we switch this to host/port? we recently moved pg to host/port, and it would be mega-nice to maintain this standard

for consistency, the config would be `host` / `port` / `username` / `password`, and the `port` could be optional and default to the DB's typical default value

Reply:
Done, thanks for the input!